File: Java/src/Type/Interval.java
package Type;

/**
 * Best-effort double-precision interval arithmetic with ULP outward padding
 * after each operation. Not IEEE-1788 rigorous (no directed-rounding mode),
 * but tight enough for diagnostic comparison against variance arithmetic.
 *
 * Conversion from VarDbl: [value - max(unc, ulp(value)), value + max(unc, ulp(value))],
 * so degenerate (unc=0) sources like IndexSin Quart still get a proper ULP-wide interval.
 */
public final class Interval {

    private final double lo;
    private final double hi;

    public Interval()                            { this.lo = 0.0; this.hi = 0.0; }
    public Interval(double point)                { this.lo = point; this.hi = point; }
    public Interval(double lo, double hi) {
        this.lo = Math.min(lo, hi);
        this.hi = Math.max(lo, hi);
    }
    public Interval(VarDbl v) {
        final double w = Math.max(v.uncertainty(), safeUlp(v.value()));
        this.lo = v.value() - w;
        this.hi = v.value() + w;
    }
    public static Interval centered(double value, double unc) {
        if (unc < 0)
            throw new IllegalArgumentException("Interval.centered: unc must be >= 0");
        final double w = Math.max(unc, safeUlp(value));
        return new Interval(value - w, value + w);
    }

    public double lo()  { return lo; }
    public double hi()  { return hi; }
    public double mid() { return 0.5 * (lo + hi); }
    public double rad() { return 0.5 * (hi - lo); }
    public boolean contains(double x)     { return lo <= x && x <= hi; }
    public boolean encloses(Interval o)   { return lo <= o.lo && o.hi <= hi; }

    public Interval negate() { return new Interval(-hi, -lo); }

    public Interval add(Interval r) {
        return padded(lo + r.lo, hi + r.hi);
    }
    public Interval sub(Interval r) {
        return padded(lo - r.hi, hi - r.lo);
    }
    public Interval mul(Interval r) {
        final double a = lo * r.lo, b = lo * r.hi;
        final double c = hi * r.lo, d = hi * r.hi;
        return padded(Math.min(Math.min(a, b), Math.min(c, d)),
                      Math.max(Math.max(a, b), Math.max(c, d)));
    }
    public Interval mul(double s) {
        final double a = lo * s, b = hi * s;
        return padded(Math.min(a, b), Math.max(a, b));
    }

    private static double safeUlp(double x) {
        final double u = Math.ulp(x);
        return (u > 0.0) ? u : Double.MIN_NORMAL;
    }
    private static Interval padded(double lo, double hi) {
        return new Interval(lo - safeUlp(lo), hi + safeUlp(hi));
    }

    @Override
    public String toString() {
        return "[" + lo + ", " + hi + "]";
    }
}

File: Python/interval.py
"""Best-effort double-precision interval arithmetic with ULP outward padding
after each operation. Not IEEE-1788 rigorous (no directed-rounding mode),
but tight enough for diagnostic comparison against variance arithmetic.

Conversion from VarDbl: [value - max(unc, ulp(value)), value + max(unc, ulp(value))],
so degenerate (unc=0) sources like IndexSin Quart still get a proper ULP-wide interval.
"""
import math
import sys
import typing

import varDbl


def _safe_ulp(x: float) -> float:
    u = math.ulp(x)
    return u if u > 0.0 else sys.float_info.min


def _padded(lo: float, hi: float) -> 'Interval':
    return Interval(lo - _safe_ulp(lo), hi + _safe_ulp(hi))


class Interval:
    __slots__ = ('_lo', '_hi')

    def __init__(self, lo: float, hi: typing.Optional[float] = None):
        if hi is None:
            hi = lo
        if lo > hi:
            lo, hi = hi, lo
        self._lo = float(lo)
        self._hi = float(hi)

    @classmethod
    def from_varDbl(cls, v: varDbl.VarDbl) -> 'Interval':
        """Half-width = max(uncertainty, ulp(value))."""
        w = max(v.uncertainty(), _safe_ulp(v.value()))
        return cls(v.value() - w, v.value() + w)

    @classmethod
    def centered(cls, value: float, unc: float) -> 'Interval':
        if unc < 0:
            raise ValueError('Interval.centered: unc must be >= 0')
        w = max(unc, _safe_ulp(value))
        return cls(value - w, value + w)

    def lo(self) -> float:   return self._lo
    def hi(self) -> float:   return self._hi
    def mid(self) -> float:  return 0.5 * (self._lo + self._hi)
    def rad(self) -> float:  return 0.5 * (self._hi - self._lo)

    def contains(self, x: float) -> bool:
        return self._lo <= x <= self._hi

    def encloses(self, other: 'Interval') -> bool:
        return self._lo <= other._lo and other._hi <= self._hi

    def __neg__(self) -> 'Interval':
        return Interval(-self._hi, -self._lo)

    def __add__(self, other):
        if isinstance(other, Interval):
            return _padded(self._lo + other._lo, self._hi + other._hi)
        return _padded(self._lo + other, self._hi + other)

    __radd__ = __add__

    def __sub__(self, other):
        if isinstance(other, Interval):
            return _padded(self._lo - other._hi, self._hi - other._lo)
        return _padded(self._lo - other, self._hi - other)

    def __rsub__(self, other):
        return _padded(other - self._hi, other - self._lo)

    def __mul__(self, other):
        if isinstance(other, Interval):
            a = self._lo * other._lo; b = self._lo * other._hi
            c = self._hi * other._lo; d = self._hi * other._hi
            return _padded(min(a, b, c, d), max(a, b, c, d))
        a = self._lo * other; b = self._hi * other
        return _padded(min(a, b), max(a, b))

    __rmul__ = __mul__

    def __repr__(self) -> str:
        return f'[{self._lo!r}, {self._hi!r}]'

    def __str__(self) -> str:
        return f'[{self._lo}, {self._hi}]'
